### PR TITLE
Reraisable exceptions

### PR DESCRIPTION
**Pull request by jsoucheiron**
Modify the way exceptions are handled by failsafe to be able to escape the failsafe for certain exceptions.

Bump version to 0.2.0

### CODE REVIEW

File: tests/test_failsafe.py [hunk 1]
```diff
@@ -108,10 +108,10 @@ def test_circuit_breaker(self):
...
 
         with pytest.raises(CircuitOpen):
-            policy = RetryPolicy(5, [SomeRetriableException])
+            policy = ExceptionHandlingPolicy(5, [SomeRetriableException])
             circuit_breaker = CircuitBreaker(maximum_failures=2)
             loop.run_until_complete(
-                Failsafe(retry_policy=policy, circuit_breaker=circuit_breaker)
+                Failsafe(exception_handling_policy=policy, circuit_breaker=circuit_breaker)
                 .run(failing_operation)
             )
```
Comment (jakubka) [inline tests/test_failsafe.py]: Can you please add tests checking that:

1) Exception was raised as expected when `abortable_exceptions` is set (similar to what we have in test_fallback_failsafe.py)
2) Failure wasn't reported to circuit breaker when abortable exception is thrown

File: failsafe/failsafe.py [hunk 2]
```diff
@@ -75,7 +77,9 @@ def __init__(self, retry_policy=None, circuit_breaker=None):
             except Exception as e:
                 context.errors += 1
                 recent_exception = e
-                retry = self.retry_policy.should_retry(context, e)
+                retry = self.exception_handling_policy.should_retry(context, e)
+                if self.exception_handling_policy.should_raise(e):
```
Comment (jsoucheiron) [inline failsafe/failsafe.py]: I'm not sure about this. If an exception is in both retriable_exceptions and abortable_exceptions what should happen?
My guess is that we should retry first and then, after the retries have been exhausted, we should raise without reporting a failure.

Comment (jakubka) [inline failsafe/failsafe.py]: Ah, good point!

I think that having exception as both retriable and abortable doesn't make sense for the user. Or does it? What would be the use case? I suggest adding logic to constructor of `RetryPolicy` which will protect from having the same exception in both lists, what do you think? That way we will have very clearly defined behaviour.

File: failsafe/failsafe.py [hunk 3]
```diff
@@ -44,8 +44,10 @@ class Failsafe:
     and the circuit breaker is always closed allowing all calls.
     """
 
-    def __init__(self, retry_policy=None, circuit_breaker=None):
-        self.retry_policy = retry_policy or RetryPolicy(0)
+    def __init__(self, exception_handling_policy=None, circuit_breaker=None):
+        if exception_handling_policy is None:
+            exception_handling_policy = ExceptionHandlingPolicy(0)
+        self.exception_handling_policy = exception_handling_policy
```
Comment (jsoucheiron) [inline failsafe/failsafe.py]: Pep8 recomends explicit is None comparison instead of bool conversion.

File: failsafe/exception_handling_policy.py [hunk 4]
```diff
@@ -11,15 +11,20 @@
...
+class ExceptionHandlingPolicy:
     """
-    Model to store the number of allowed retries
-    and the allowed retriable exceptions.
+    Model to store the number of allowed retries, the allowed retriable exceptions
+    and the exceptions that should be re-raised
     """
 
-    def __init__(self, allowed_retries=3, retriable_exceptions=None):
+    def __init__(self, allowed_retries=3, retriable_exceptions=None, raisable_exceptions=None):
```
Comment (jsoucheiron) [inline failsafe/exception_handling_policy.py]: OK

File: failsafe/exception_handling_policy.py [hunk 5]
```diff
@@ -34,3 +39,14 @@ def should_retry(self, context, exception=None):
 
     def _is_expected_exception(self, exception):
```
Comment (jsoucheiron) [inline failsafe/exception_handling_policy.py]: OK

File: failsafe/exception_handling_policy.py [hunk 6]
```diff
@@ -34,3 +39,14 @@ def should_retry(self, context, exception=None):
...
         return any(isinstance(exception, e) for e in self.retriable_exceptions) if self.retriable_exceptions else True
+
+    def should_raise(self, exception=None):
```
Comment (jsoucheiron) [inline failsafe/exception_handling_policy.py]: OK

File: tests/test_failsafe.py (hunk 1 above)
Comment (jsoucheiron) [inline tests/test_failsafe.py]: Done

**Comment by jakubka** [comment]
@jsoucheiron commit with rename + few others pushed. Can you please review?

I am not convinced about the last commit with logging statements...

File: README.md [hunk 7]
```diff
@@ -88,15 +88,16 @@ RetryPolicy instances are stateless. They can be safely shared between Failsafe
...
 
 ```python
 from failsafe import Failsafe, RetryPolicy
 
 async def my_async_function():
-    raise ValueError()  # by default, every exception will cause a retry
+    raise ValueError()  # ValueError is an abortable exception, so it will not cause retry
 
-retry_policy = RetryPolicy(abortable_exceptions=[ValueError])
+retry_policy = RetryPolicy(allowed_retries=4, abortable_exceptions=[ValueError])
```
Comment (jsoucheiron) [inline README.md]: allowed retries will be ignored

File: failsafe/retry_policy.py [hunk 8]
```diff
@@ -37,15 +42,21 @@ def should_retry(self, context, exception=None):
...
             abortable or not during the execution.
         """
-        return any(isinstance(exception, e) for e in self.abortable_exceptions) if self.abortable_exceptions else False
+        if self.abortable_exceptions is None:
+            return False
+
+        return any(isinstance(exception, e) for e in self.abortable_exceptions)
+
+    def _is_retriable_exception(self, exception):
+        if self.retriable_exceptions is None:
```
Comment (jsoucheiron) [inline failsafe/retry_policy.py]: This will never happen. In this case I'd just remove the is None.

File: failsafe/retry_policy.py [hunk 9]
```diff
@@ -37,15 +42,21 @@ def should_retry(self, context, exception=None):
...
-        Returns a boolean indicating if we should abort the run and propagate the given exception
-        outside the failsafe
+        Returns a boolean indicating whether failsafe run should be aborted and exception propagated
+        outside the failsafe.
 
         :param exception: Exception which caused failure to be considered
             abortable or not during the execution.
         """
-        return any(isinstance(exception, e) for e in self.abortable_exceptions) if self.abortable_exceptions else False
+        if self.abortable_exceptions is None:
```
Comment (jsoucheiron) [inline failsafe/retry_policy.py]: Same here

File: README.md (hunk 7 above)
Comment (jakubka) [inline README.md]: Yeah, I wanted to make it obvious that there are some retries allowed, but they will not be executed. Maybe this was not necessary...

**Pull request merged**